File: code/scenes/src/main/java/org/fbme/scenes/controllers/SelectionModelBase.java
package org.fbme.scenes.controllers;

import org.jetbrains.annotations.NotNull;

import java.util.HashSet;
import java.util.Set;

public abstract class SelectionModelBase<T> implements SelectionModel<T> {

    private final Set<SelectionModel.Listener<T>> myListeners = new HashSet<>();

    protected final void fireSelected(@NotNull T component, boolean selected) {
        for (SelectionModel.Listener<T> listener : myListeners) {
            listener.onComponentSelected(component, selected);
        }
    }

    @Override
    public final void addListener(@NotNull SelectionModel.Listener<T> listener) {
        myListeners.add(listener);
    }

    @Override
    public final void removeListener(@NotNull SelectionModel.Listener<T> listener) {
        myListeners.remove(listener);
    }

}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/SelectionModelBase.kt
package org.fbme.scenes.controllers

abstract class SelectionModelBase<T> : SelectionModel<T> {
    private val listeners: MutableSet<SelectionModel.Listener<T>> = HashSet()

    protected fun fireSelected(component: T, selected: Boolean) {
        for (listener in listeners) {
            listener.onComponentSelected(component, selected)
        }
    }

    override fun addListener(listener: SelectionModel.Listener<T>) {
        listeners.add(listener)
    }

    override fun removeListener(listener: SelectionModel.Listener<T>) {
        listeners.remove(listener)
    }
}
